### PR TITLE
VQVAE Model for HipMRI Study Image Reconstruction

### PR DESCRIPTION
**Pull request by carl-c2**
This pull request adds the VQVAE project for medical image reconstruction. The project includes:

- Implementation of VQVAE in `modules.py`.
- Data preprocessing and loading in `dataset.py`.
- Training and validation of model, alongside output of results in `train.py`.
- Prediction and visualizations in `predict.py`.
- Complete documentation in `README.md`.

Carl Chang - s4748941

### CODE REVIEW

**Comment by Uto0op** [comment]
Recognition problem: VQVAE on HipMRI ---- Hard Difficulty

1. Algorithm solves the problem:
Very clear and resembling reconstructed images and SSIM over 0.7 (5 Marks)

2. Algorithm demonstrated and functions: (3 Marks)

3. Design:
Modularized design and good programming practice using `utils.py` (1 Mark)

4. Commenting:
Meaningful comments (1 Mark)

5. Algorithm above Normal Difficulty: (5 Marks)

6. Algorithm is Hard difficulty: (5 Marks)

Suggestions:

- You could include the hyperparameter chosen within the training process of README.md, as well as justification on the hyperparameter tuning process (e.g. why you ended up choosing LR of 1e-3, what else you had attempted to improve the performance and what were their implications);
- Good metric plots but no (brief) explanation on each of them.

**Comment by YashMitttal** [comment]
Some additional areas of improvement:

- While you've shown both original and reconstructed images, a brief analysis of where the model performs well or struggles (e.g., which image regions or features are less accurately reconstructed) would be good to write.

- Since SSIM is a perceptual metric, including a short explanation of why your SSIM score is considered acceptable for your data would add context.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. VQVAE (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Acceptable. Good metric plots but no (brief) explanation.</td><td>-1</td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF submitted.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time submission.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-2</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Great work!
